qiskit/dagcircuit/dagdependency_v2.py

```diff
@@ -0,0 +1,740 @@
+        max_node = self.get_node(max_node_id)
+
+        reachable = [True] * max_node_id
+
+        # Analyze nodes in the reverse topological order.
+        # An improvement to the original algorithm is to consider only direct predecessors
+        # and to avoid constructing the lists of forward and backward reachable predecessors
+        # for every node when not required.
+        for prev_node_id in range(max_node_id - 1, -1, -1):
+            if reachable[prev_node_id]:
+                prev_node = self.get_node(prev_node_id)
+
+                if not self.comm_checker.commute(
+                    prev_node.op,
+                    prev_node.qargs,
+                    prev_node.cargs,
+                    max_node.op,
+                    max_node.qargs,
+                    max_node.cargs,
+                ):
+                    # If prev_node and max_node do not commute, then we add an edge
+                    # between the two, and mark all direct predecessors of prev_node
+                    # as not reaching max_node.
+                    self._multi_graph.add_edge(prev_node_id, max_node_id, {"commute": False})
+
+                    predecessor_ids = self.predecessor_indices(prev_node_id)
+                    for predecessor_id in predecessor_ids:
+                        reachable[predecessor_id] = False
+            else:
+                # If prev_node cannot reach max_node, then none of its predecessors can
+                # reach max_node either.
+                predecessor_ids = self.predecessor_indices(prev_node_id)
+                for predecessor_id in predecessor_ids:
+                    reachable[predecessor_id] = False
+
+    def get_node(self, node_id):
+        """
+        Args:
+            node_id (int): label of considered node.
+
+        Returns:
+            node: corresponding to the label.
+        """
+        return self._multi_graph.get_node_data(node_id)
+
+    def remove_all_ops_named(self, opname):
+        """Remove all operation nodes with the given name."""
+        for n in self.named_nodes(opname):
+            self.remove_op_node(n)
+
+    def named_nodes(self, *names):
+        """Get the set of "op" nodes with the given name."""
+        named_nodes = []
+        for node in self._multi_graph.nodes():
+            if isinstance(node, DAGOpNode) and node.op.name in names:
+                named_nodes.append(node)
+        return named_nodes
+
+    def _increment_op(self, op):
+        if op.name in self._op_names:
+            self._op_names[op.name] += 1
+        else:
+            self._op_names[op.name] = 1
+
+    def _decrement_op(self, op):
+        if self._op_names[op.name] == 1:
+            del self._op_names[op.name]
+        else:
+            self._op_names[op.name] -= 1
+
+    def count_ops(self):
+        """Count the occurrences of operation names."""
+        return self._op_names
+
+    def remove_op_node(self, node):
+        """Remove an operation node n.
+
```

Comment:
I am a little bit worried about this function in the context of DAG dependency. Though (after some thinking) I do believe that this method of removing nodes and reconnecting predecessors to successors is safe: we will not be able to incorrectly infer from dag dependency that two nodes commute when in fact they do not. It might not be compete: the dag dependency may say that two nodes do not commute when in fact they do. For example, suppose that we have a chain of nodes `A -> B -> C`. By definition, `A` and `B` do not commute, and `B` and `C` do not commute (but this says nothing about whether `A` and `C` commute). If we remove `B` and retain edges, we will end up with `A -> C`, which might be over-conservative if `A` and `C` do commute. Note that it's not enough to check the commutativity of `A` and `C` and remove the edge if they do commute, since we should also recursively check if `C` commutes with predecessors of `A`, and `A` commutes with successors of `C`. So at the minimum let's update the docstring of this function. Hmm, when would we ever want to remove nodes from a DAG?

Reply:
The comment for `replace_block_with_op` addresses the question of commutativity of the output, which is relevant here as well. `remove_op_node` is used extensively in the transpiler passes.